### PR TITLE
Fixing Issue #26:  Unexpected EOF when backend connection is closed

### LINKED ISSUE #26: server read err: unexpected EOF
Randomly, I'm having requests return a "Bad Gateway" error.  It's due to the following error getting thrown by strowger:

```
server read err: unexpected EOF
```

Error is thrown [here](https://github.com/flynn/strowger/blob/faf0cfcdd0a325403e74e6e1bb2dff3e68bdc1cb/http.go#L427).

### PR DESCRIPTION
**Pull request by jayd3e**
Closes #26.

### CODE REVIEW

File: http_test.go [hunk 1]
```diff
@@ -171,3 +171,25 @@ func (s *S) TestHTTPInitialSync(c *C) {
...
+	defer discoverd.UnregisterAll()
+
+	addHTTPRoute(c, l)
+
+	// a single request is allowed to successfully get issued
+	assertGet(c, "http://"+l.Addr, "example.com", "1")
+
+	// the backend server's connection gets closed, but the server
+	// is able to recover
+	srv.Close()
```
Comment (titanous) [inline http_test.go]: This closes the whole server down, not just the connection.

Comment (titanous) [inline http_test.go]: You want `srv.CloseClientConnections()`

Comment (jayd3e) [inline http_test.go]: Ok cool, sounds good.

**Comment by jayd3e** [comment]
Just to give some context to where this is at.  I'm attempting to use the `conn.Read([]byte)` method to determine if our connection has been closed or not before writing.  Problem is, the connection seems to be getting reset(in order words con.Read() is returning io.EOF) more often that expected.  After looking into it a little, I think Go might squash a number of linux level errors into io.EOF, but I'm not certain of this.

File: http.go [hunk 2]
```diff
@@ -373,6 +373,7 @@ func (s *httpService) handle(req *http.Request, sc *httputil.ServerConn, tls boo
 	}
 	defer backend.Close()
 
+	var zero []byte
```
Comment (titanous) [inline http.go]: We didn't end up using this, right? Do we also need the above changes passing the `conn` around?

File: http.go [hunk 3]
```diff
@@ -431,6 +432,10 @@ func (s *httpService) handle(req *http.Request, sc *httputil.ServerConn, tls boo
 			return
 		}
 
+		// close the backend connection, so we don't accidently send to
+		// a closed socket on the backend
+		backend.Close()
```
Comment (titanous) [inline http.go:435]: We also need to get a new connection after reading a new request.

File: http.go (hunk 2 above)
Comment (jayd3e) [inline http.go]: Yup you're right.

File: http.go (hunk 3 above)
Comment (jayd3e) [inline http.go:435]: Hmmm I'm curious as to how the tests passed without refreshing the connection.  Locally that is, it seems Travis has other issues.

File: http.go (hunk 2 above)
Comment (jayd3e) [inline http.go]: Right on both accounts.

File: http.go (hunk 3 above)
Comment (titanous) [inline http.go:435]: Yeah, that needs sorting out. net/http probably opened a new connection for some reason.

**Comment by titanous** [comment]
LGTM. Please squash and add a [DCO sign-off](https://flynn.io/docs/contributing#developers-certificate-of-origin) and then I'll merge.

**Comment by jayd3e** [comment]
OK awesome, will do first thing tomorrow.

**Comment by jayd3e** [comment]
Good to go.

**Comment by titanous** [comment]
Thanks!

**Pull request merged**